### PR TITLE
Create Search Page

### LINKED ISSUE #59: Edit Navigation Bar
1. For non-sign-in users or non-registered users only show register and login options.
2. For sign-in users users show All-jobs, logout, search options.
   2.1 For student user show Profile option.
   2.2 For student user show Add-Job option.

### LINKED ISSUE #61: Create Search Page
Filtering jobs by the search cell in the Navigation bar (for now only by title) .

### PR DESCRIPTION
**Pull request by shayakrach**
Filtering jobs by the search cell in the Navigation bar by title.
Only registered (HR's and students) and log-in users can use
the search feature. This search feature increases user satisfaction
by improving the usability, accessibility, and efficiency of user
interaction with the website.

Close #61 

I recognized a typo that I made in a previous PR.
This typo will be fixed in another PR that closes issue #71.

### CODE REVIEW

**Review by AshkenYariv — APPROVED** [review]
Looks good!

**Review by assafg95 — APPROVED** [review]

**Comment by assafg95** [comment]
Looks good to me too

File: templates/base.html [hunk 1]
```diff
@@ -38,8 +38,8 @@
               <a class="nav-link" href="#">Profile</a>
             </li>
             {% endif %}
-            <form class="d-flex">
-              <input class="form-control me-2" type="search" placeholder="Search" aria-label="Search">
+            <form class="d-flex" action="{% url 'jobs_search' %}" method="get">
+              <input name="q" class="form-control me-2" type="search" placeholder="Search" aria-label="Search...">
```
Comment (ormergi) [inline templates/base.html:42]: What is `q`? please rename it to more informative name.

**Review by ormergi — COMMENTED** [review]
first commit:
Why this commit is necessary for this PR?
If it can be done on another PR please move it.

File: templates/base.html (hunk 1 above)
Comment (ormergi) [inline templates/base.html]: If its not something that related to Django, please rename `q` to more informative (e.g: search_query).

**Review by ormergi — CHANGES_REQUESTED** [review]
second commit:
Please add tests for the new business logic (**at least**: search actually return what you expect, redirection to search page works).
Add them to this commit and not on another one.

**Comment by ormergi** [comment]
@shayakrach  ping

**Comment by shayakrach** [comment]
> first commit: Why this commit is necessary for this PR? If it can be done on another PR please move it.

It is not.
Moved to #77.

**Comment by shayakrach** [comment]
> @shayakrach ping

Hi @ormergi,

This PR is in WIP status right now.

When it will be ready I will post it on Slack,

**Comment by shayakrach** [comment]
> first commit: Why this commit is necessary for this PR? If it can be done on another PR please move it.

Already moved.

**Comment by shayakrach** [comment]
> second commit: Please add tests for the new business logic (**at least**: search actually return what you expect, redirection to search page works). Add them to this commit and not on another one.

Fixed @ormergi

File: conftest.py [hunk 2]
```diff
@@ -97,5 +97,33 @@ def job_2(user_HR):
                          post_until='2022-12-23',
                          is_active=True,
                          marked_count=0,
-                         apply_link=DEFAULT_APPLY_LINK)
+                         apply_link=DEFAULT_APPLY_LINK+'2')
```
Comment (ormergi) [inline conftest.py]: Why this change is here?

File: conftest.py [hunk 3]
```diff
@@ -97,5 +97,33 @@ def job_2(user_HR):
...
+    return job
+
+
+@pytest.fixture
+def job_SoftwareEngineer(user_HR):
```
Comment (ormergi) [inline conftest.py]: This function name does not follow [python function name convention ](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names)
and its not consistent with the reset of the project.
Please change it and other similar function names.

Also, it seems that you use the new fixtures only for the search page tests, please move them to the same file.

File: conftest.py [hunk 4]
```diff
@@ -97,5 +97,33 @@ def job_2(user_HR):
...
+                         company_name=DEFAULT_COMPANY_NAME,
+                         post_until='2022-12-23',
+                         is_active=True,
+                         marked_count=0,
+                         apply_link=DEFAULT_APPLY_LINK+'3')
+    return job
+
+
+@pytest.fixture
+def job_DataEngineer(user_HR):
```
Comment (ormergi) [inline conftest.py]: Same

File: jobs/tests/test_jobs_view.py [hunk 5]
```diff
@@ -30,6 +30,7 @@ def test_unauthenticated_user_enter_register(client):
 
 
 # Should move move another file - Good for Demo 3
+# Fix the typo here - issue #71
```
Comment (ormergi) [inline jobs/tests/test_jobs_view.py]: There is already an issue open for this, there is no need to add a comment as well especially on  non related PR.

File: jobs/tests/test_search_jobs_view.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
+import pytest
+from jobs.models import Job
+
+
+@pytest.mark.django_db
+def test_search_page_filtering_by_title_successfully_by_any_partial_title(client, user_HR, job_DataEngineer):
```
Comment (ormergi) [inline jobs/tests/test_search_jobs_view.py]: I would remove `..search_page..` part from all the fixture names,
their names already long and it could be understood by the file name.
Something like `test_search_by_partial_job_title`

File: jobs/tests/test_search_jobs_view.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    assert Job.objects.count() == 1
```
Comment (ormergi) [inline jobs/tests/test_search_jobs_view.py]: I think this assertion should be part of `job_DataEngineer` fixture

File: jobs/tests/test_search_jobs_view.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        assert len(filtered_jobs) == 1
+        assert set(filtered_jobs) == set(Job.objects.filter(title__icontains=partial_title))
+
+
+@pytest.mark.django_db
+def test_search_page_filtering_by_title_with_bad_title(client, user_HR, job_DataEngineer):
+    assert Job.objects.count() == 1
+
+    client.force_login(user_HR)
+    title = job_DataEngineer.title + 'e'
```
Comment (ormergi) [inline jobs/tests/test_search_jobs_view.py]: This one deserves its own fixture instead of manipulation one that exists.

**Review by ormergi — CHANGES_REQUESTED** [review]
Hi Shay, 

I started to review this PR starting for reading its description then reviewed the the tests in order to understand how it works.
From what I understand this feature is targeted only for HR users,
if this is not the case I would expect to see various users types on tests.
Also I would expect to see at least tree elements on the job page when you test the search feature.

The PR description should also include motivation for the feature (how its improving the app e.g: UX improvement), who can use it (HR users only?), same goes for the commit message.

**Comment by shayakrach** [comment]
> 



> Hi Shay,
> 
> I started to review this PR starting for reading its description then reviewed the the tests in order to understand how it works. From what I understand this feature is targeted only for HR users, if this is not the case I would expect to see various users types on tests. Also I would expect to see at least tree elements on the job page when you test the search feature.
> 
> The PR description should also include motivation for the feature (how its improving the app e.g: UX improvement), who can use it (HR users only?), same goes for the commit message.



> Hi Shay,
> 
> I started to review this PR starting for reading its description then reviewed the the tests in order to understand how it works. From what I understand this feature is targeted only for HR users, if this is not the case I would expect to see various users types on tests. Also I would expect to see at least tree elements on the job page when you test the search feature.
> 
> The PR description should also include motivation for the feature (how its improving the app e.g: UX improvement), who can use it (HR users only?), same goes for the commit message.

I made all the changes you asked for @ormergi .
I think it is better now.

**Review by assafg95 — APPROVED** [review]
Everything Looks Great!

File: jobs/tests/test_search_jobs_view.py [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+
+NUM_OF_JOBS = 10
+
+
+@pytest.fixture
+def software_engineer_jobs(user_HR):
+    software_engineer_jobs = []
+    for i in range(NUM_OF_JOBS):
+        software_engineer_job = Job.create_job(
+                                publisher=user_HR, title='Software Engineer',
```
Comment (ormergi) [inline jobs/tests/test_search_jobs_view.py]: nit: move title to new line.

File: jobs/tests/test_search_jobs_view.py [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+        software_engineer_jobs.append(software_engineer_job)
+    return software_engineer_jobs
+
+
+@pytest.fixture
+def data_engineer_jobs(user_HR):
+    data_engineer_jobs = []
+    for i in range(NUM_OF_JOBS):
+        data_engineer_job = Job.create_job(
+                            publisher=user_HR, title='Data Engineer',
```
Comment (ormergi) [inline jobs/tests/test_search_jobs_view.py]: nit: move title to new line.

**Review by ormergi — CHANGES_REQUESTED** [review]
Hi Shay,

Please change the vars and function names according to python naming convention, they arr also apply for acronyms (HR -> hr).

Regarding tests fixture names, try to rephrase them to be more like natural language, for example `test_search_by_partial_job_title_by_HR` -> `test_hr_user_job_search_by_title`

Note about tests, in general its a good practice to treat tests as the documentation of your project,
it allows new comers to understand the flow, how it works and how to use it (in case it is a library).

**Comment by shayakrach** [comment]
> Hi Shay,
> 
> Please change the vars and function names according to python naming convention, they arr also apply for acronyms (HR -> hr).
> 
> Regarding tests fixture names, try to rephrase them to be more like natural language, for example `test_search_by_partial_job_title_by_HR` -> `test_hr_user_job_search_by_title`
> 
> Note about tests, in general its a good practice to treat tests as the documentation of your project, it allows new comers to understand the flow, how it works and how to use it (in case it is a library).

Fixed @ormergi

**Review by ormergi — APPROVED** [review]
Thanks!
